httpcore/_async/http11.py

```diff
@@ -77,6 +79,20 @@ async def handle_async_request(self, request: Request) -> Response:
             )
 
         async with self._state_lock:
+            if self._state == HTTPConnectionState.SERVER_DISCONNECTED:
+                raise ServerDisconnectedError()
+
+            # If the HTTP connection is idle but the socket is readable, then the
+            # only valid state is that the socket is about to return b"", indicating
+            # a server-initiated disconnect.
+            server_disconnected = (
+                self._state == HTTPConnectionState.IDLE
+                and self._network_stream.get_extra_info("is_readable")
+            )
+            if server_disconnected:
+                self._state = HTTPConnectionState.SERVER_DISCONNECTED
+                raise ServerDisconnectedError()
+
```

Comment:
most of users make subclass of `handle_async_request` to customize their _Client_ instances. I think it could be better for us to keep it cleaner and more readable. 
Could you export this change into priavte method `_raise_for_state` which have no returns and only raises if possible.

Reply:
Done, moved it into `_update_state` function as we also need to update the state as part of raising. (Also for http2 to have it consistent)